mesonbuild/modules/_qt.py

```diff
@@ -616,3 +733,431 @@ def compile_translations(self, state: 'ModuleState', args: T.Tuple, kwargs: 'Com
+            install_tag.append('devel')
+
+        if collected_json:
+            cmd.append('@INPUT@')
+
+        return build.CustomTarget(
+            f'typeregistrar_{target_name}',
+            state.subdir,
+            state.subproject,
+            state.environment,
+            cmd,
+            inputs,
+            outputs,
+            install=kwargs['install'],
+            install_dir=install_dir,
+            install_tag=install_tag,
+            description=f'Qml type registration for {target_name}',
+        )
+
+    @FeatureNew('qt.qml_module', '1.6')
+    @typed_kwargs(
+        'qt.qml_module',
+        #qml sources
+        KwargInfo('qml_sources', ContainerTypeInfo(list, (File, str, build.CustomTarget)), listify=True, default=[]),
+        KwargInfo('qml_singletons', ContainerTypeInfo(list, (File, str, build.CustomTarget)), listify=True, default=[]),
+        KwargInfo('qml_internals', ContainerTypeInfo(list, (File, str, build.CustomTarget)), listify=True, default=[]),
+        KwargInfo('resources_prefix', str, default='qt/qml'),
+
+        #qmldir generation
+        KwargInfo('imports', ContainerTypeInfo(list, (str)), default=[]),
+        KwargInfo('optional_imports', ContainerTypeInfo(list, (str)), default=[]),
+        KwargInfo('default_imports', ContainerTypeInfo(list, (str)), default=[]),
+        #match DEPENDENCIES argument from CMake, but dependencies keyword is already taken
+        KwargInfo('depends_imports', ContainerTypeInfo(list, (str)), default=[]),
+        KwargInfo('designer_supported', bool, default=False),
+
+        #for type registration, same arguments as moc
+        KwargInfo('moc_headers', ContainerTypeInfo(list, (File, str, build.CustomTarget)), listify=True, default=[]),
+        KwargInfo('moc_sources', ContainerTypeInfo(list, (File, str, build.CustomTarget)), listify=True, default=[]),
+        KwargInfo('include_directories', ContainerTypeInfo(list, (build.IncludeDirs, str)), listify=True, default=[]),
+        KwargInfo('namespace', str, default=''),
+        KwargInfo('typeinfo', str, default=''),
+
+        KwargInfo('moc_extra_args', ContainerTypeInfo(list, str), listify=True, default=[]),
+        KwargInfo('qmlcachegen_extra_args', ContainerTypeInfo(list, str), listify=True, default=[]),
+        KwargInfo('qmltyperegistrar_extra_args', ContainerTypeInfo(list, str), listify=True, default=[]),
+
+        KwargInfo('generate_qmldir', bool, default=True),
+        KwargInfo('generate_qmltype', bool, default=True),
+        KwargInfo('cachegen', bool, default=True),
+
+        KwargInfo('dependencies', ContainerTypeInfo(list, (Dependency, ExternalLibrary)), listify=True, default=[]),
+        INSTALL_DIR_KW,
+        INSTALL_KW,
+        KwargInfo('method', str, default='auto'),
+        KwargInfo('preserve_paths', bool, default=False),
+    )
+    def qml_module(self, state: ModuleState, args: T.Tuple[str, str, str], kwargs: 'QmlModuleKwArgs') -> ModuleReturnValue:
```

Comment:
The type of `args` is not correct. Nothing in Meson prevents people from calling it with more or fewer arguments. It should be defined as a list of strings. It should be a `T.List[str]`. The code below should also check that its size is exactly 3 and that every entry is actually a string and error out if they are not.

Reply:
Well, normally we'd do that with typed_pos_args, surely. :)

Reply:
I went for `typed_pos_args`